### ISSUE TITLE
Wrong interpretation of `let-foo` on `<template>`

### ISSUE THREAD

**Issue by zoechi** [dart-lang/angular_analyzer_plugin#321]
![image](https://cloud.githubusercontent.com/assets/405837/26053824/994522ae-396a-11e7-98a6-891ebd0a8a07.png)

The error message indicates that `let-item` is interpreted like what `#item` is supposed to do - to create a reference to the `<template>` element, but `let-item` is supposed to provide a reference to the `$implicit` context property passed to `createEmbeddedView` (where the component is stamped), like `item` in `*ngFor="let item of items"`.

Somewhat related to 

- #167
- #176

**Comment by MichaelRFairhurst** [comment]
Hey again Gunter!

Interesting. I thought this was actually how it worked, I thought that html references and components were provided in `$implicit` and therefore bound via `#foo`.

You can see my source of confusion in the tests: [#x test](https://github.com/dart-lang/angular2/blob/9eb1ef80eb09ac72c0e24e0ea3a6dcd46d036544/test/compiler/expression_parser/parser_test.dart#L396) and [let x test](https://github.com/dart-lang/angular2/blob/9eb1ef80eb09ac72c0e24e0ea3a6dcd46d036544/test/compiler/expression_parser/parser_test.dart#L403), but this is only for inside a template binding. Outside of a template binding `#foo` is a `ReferenceAst`, and its handled completely differently.

This should be fixed: we should stop treating all references as `$implicit`, and not treat let bound vs hash bound the same way.

**Comment by zoechi** [comment]
```html
<template #foo let-item let-bar="bar">
  <div>{{item}}</div>
  <div>{{bar}}</div>
</template>
<div [myDirective] [templateRef]="foo"></div>
```
```dart
@Directive(selector: '[myDirective]')
class MyDirective {
  final ViewContainerRef _viewContainerRef;
  NgTemplateOutlet(this._viewContainerRef);

  @Input()
  TemplateRef templateRef

  stampTemplate() {
    _insertedViewRef = _viewContainerRef.createEmbeddedView(templateRef);
    _insertedViewRef.setLocal(r'$implicit', 'implicitContent');
    _insertedViewRef.setLocal('bar', 'barContent');
  }
}
```
will result in 
```html
<div>implicitContent</div>
<div>bar</div>
```
- `#foo` introduces a template variable that reference the template element
- `let-item` introduces a context variable inside the template that references the item added with key `$implicit` to the context 
- `let-bar="bar"` introduces a context variable inside the template that references the item added with key `bar` to the context

I hope this makes it more clear.

**Comment by mk13** [comment]
Looking at this, I'm not sure if we can grab the type of whatever $implicit is tied to. In such a case, maybe just return a dynamic type to it? 

However, for *ngFor, we can still link item in 'let item of items' to the runtime type of 'items'.

**Comment by zoechi** [comment]
@mk13 that's right, there is no way of knowing what type `$implicit` or `bar` will have. The template can be stamped multiple times with different values. Angular would need some additional construct to allow to define the type (or some comment syntax the plugin can interpret).
I also think using `dynamic` is the best way to tackle this (at least in the short term).

For `*ngFor` it's implicit knowledge because we know what `*ngFor` is doing.

**Comment by mk13** [comment]
Should be fixed with PR #348 . Reopen if problem persists.
